modules/plugin/geotiff/src/test/java/org/geotools/gce/geotiff/GeoTiffCRSFactoryDeadlockTest.java

```diff
@@ -0,0 +1,90 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ *
+ *    (C) 2021, Open Source Geospatial Foundation (OSGeo)
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ *
+ */
+package org.geotools.gce.geotiff;
+
+import static java.lang.Thread.sleep;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.measure.Unit;
+import org.geotools.referencing.CRS;
+import org.geotools.referencing.factory.AllAuthoritiesFactory;
+import org.geotools.util.factory.Hints;
+import org.junit.Test;
+import org.opengis.referencing.FactoryException;
+import org.opengis.referencing.crs.CRSAuthorityFactory;
+
+public class GeoTiffCRSFactoryDeadlockTest {
+
+    @Test
+    public void testDeadlockOnCRSFactories() throws ExecutionException, InterruptedException {
+        boolean potentialDeadlock = false;
+        final long timeout = 5;
+
+        final boolean[] exceptionOccurred = {false};
+        ExecutorService executorService = Executors.newFixedThreadPool(2);
+        Future[] futures = new Future[2];
```

Comment:
Did the test actually lock up frequently before the fix? 
Using just 2 calls is the lightest possible load, I was expecting to see a storm of tens of parallel executors all thrown together on a thread pool with 8-16 threads to make this lock up.

Reply:
Before the fix, the test was failing very often (I would say 8 of 10 runs).
I was also testing it in a loop and with more threads but it didn't change very much.
I'll double check and retry.

Reply:
Done